AWS-CodePipeline/sam-monorepo-pipeline-template/{{cookiecutter.outputDir}}/pipeline/pipeline_trigger.py

```diff
@@ -0,0 +1,75 @@
+import os, boto3
+from aws_lambda_powertools.logging import Logger
+from aws_lambda_powertools.utilities import parameters
+
+logger = Logger()
+cc = boto3.client('codecommit')
+cp = boto3.client('codepipeline')
+ssm = boto3.client('ssm')
+
+ssm_prefix = os.environ.get('SSM_PREFIX', '')
+app_dir = os.environ.get('APP_DIR_IN_MONOREPO')
+pipeline_name = os.environ.get('PIPELINE_NAME')
+
+@logger.inject_lambda_context
+def lambda_handler(event, _):
+    logger.info(event)
+    commit_id = event['detail']['commitId']
+    branch_name = event['detail']['referenceName']
+    repository = event['detail']['repositoryName']
+
+    paths = get_modified_files_since_last_run(repository, commit_id, branch_name)
+    toplevel_dirs = get_unique_toplevel_dirs(paths)
+    logger.info({'paths': paths, 'toplevel_dirs': toplevel_dirs, 'pipeline_name': pipeline_name})
+
+    if app_dir in toplevel_dirs:
+        start_codepipeline(pipeline_name)
+    else:
+        logger.info('Not triggering Pipeline %s. No changes under App dir %s', pipeline_name, app_dir)
+
+    update_last_commit(repository, commit_id, branch_name)
+
+def get_unique_toplevel_dirs(modified_files):
+    toplevel_dirs = set(filter(lambda a: len(a) > 1, [file.split('/')[0] for file in modified_files]))
+    logger.info('toplevel dirs: %s', toplevel_dirs)
+    return list(toplevel_dirs)
+
+def start_codepipeline(codepipeline_name):
+    try:
+        cp.start_pipeline_execution(name=codepipeline_name)
+        logger.info(f'Started CodePipeline {codepipeline_name}.')
+    except cp.exceptions.PipelineNotFoundException:
+        logger.info(f'Could not find CodePipeline {codepipeline_name}.')
+        return False
+    return True
+
+def param_name_last_commit(repository, branch_name):
+    return os.path.join('/', ssm_prefix, repository, branch_name, app_dir, 'LastCommit')  # type: ignore
+
+def get_last_commit(repository, commit_id, branch_name):
```

Comment:
Would this have race condition? Imagine there are many sub folders, so codecommit will trigger many lambda functions at the same. If one of them call the `update_last_commit` before some calling `get_last_commit`. What will happen?

Reply:
Not really, because the SSM path here includes `app_dir`, which is unique to each Lambda.

Reply:
Got it!